illwill.nim

```diff
@@ -1617,4 +895,230 @@ proc drawRect*(tb: var TerminalBuffer, x1, y1, x2, y2: Natural,
   ## Convenience method to draw a rectangle into a terminal buffer directly.
   var bb = newBoxBuffer(tb.width, tb.height)
   bb.drawRect(x1, y1, x2, y2, doubleStyle)
-  tb.write(bb)
\ No newline at end of file
+  tb.write(bb)
+
+var
+  userInputText = ""
+  userInputActive = false
+  userInputExitKey = Key.None
+  userInputMaxLen = 1
+  userInputColumn = 0
+  userInputRow = 0
+  userInputAllowTabArrow = false
+
+proc startInputLine*(
+                      tb: var TerminalBuffer, 
+                      x: int, 
+                      y: int, 
+                      strDefault: string, 
+                      strLen: int, 
+                      allowTabArrowExit = false
+                    ) =
+  ## Setup the terminal for gathering text from the terminal user.
+  ##
+  ## The ``strDefault`` value is displayed at the column (x) and row(y). The
+  ## cursor is then moved to the end of that value and shown.
+  ##
+  ## If the column is too close the right edge fo the screen, then the length
+  ## is adjusted to account for that.
+  ##
+  ## This procedure does NOT actually gather then number. Instead, future calls
+  ## to ``inputLineReady`` will process future keypresses to build the string
+  ## and determine if the string is ready to be used.
+  ##
+  ## By default, only Enter (on Windows/POSIX) will finish the string entry (or
+  ## Ctrl-Enter on Javascript.) However, if ``allowTabArrowExit`` is set to true
+  ## then Escape, Up, Down, and Tab will also finish the entry.
```

Comment:
The list of exit keys seems arbitrary; because you are putting this into a generic library you need to make it generic enough so it's flexible enough for different use cases. I'd change this so it accepts a sequence of exit keys instead of just a flag (no need to validate it, if the user passes in stupid stuff like the keys A, B and C, it's their problem...)